Add annotation to disable "https" or "clientserver" TLS auth

kcmackeen opened the pull request:
Add a new annotation:

```
vertica.com/disable-tls-auth-for-config
```

Which accepts the values "https" or "clientserver". This allows us to turn on TLS auth (using ```vertica.com/enable-tls-auth```) but disable it for one of the two TLS configs.

LiboYu2 commented:
Why do we only support disabling one of them? What if the customers want to disable both of them?

kcmackeen commented:
> > Add a new annotation:
> > ```
> > vertica.com/disable-tls-auth-for-config
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Which accepts the values "https" or "clientserver". This allows us to turn on TLS auth (using `vertica.com/enable-tls-auth`) but disable it for one of the two TLS configs.
> 
> I looked at your change. The change only stops the operator from using TLS. The TLS configuration for the vertica DB is still here. I thought your change would revert all the changes that are applyed when enable-tls-auth is turned on. This is more about stopping using TLS in operator.

Great point; will be addressed in a future PR

roypaulin commented on api/v1/helpers.go:
```diff
@@ -2211,12 +2248,22 @@ func (v *VerticaDB) GetClientServerTLSSecret() string {
+// 2) DB not initialized
+// 3) Rotate has failed (and rollback is not in progress)
+// 4) TLS auth is disabled for this config
+func (v *VerticaDB) ShouldSkipTLSUpdateReconcile(tlsConfigName string, isRollback bool) bool {
```
You could simplify this by having a helper function and then have one version for each tls config.

roypaulin commented on pkg/controllers/vdb/tlsconfig_reconciler.go:
```diff
@@ -155,3 +164,10 @@ func (h *TLSConfigReconciler) checkIfTLSAuthenticationCreatedInDB(ctx context.Co
 	return res == "True", nil
 }
+
+func (h *TLSConfigReconciler) shouldSkipTLSConfigReconcile() bool {
```
A detailed comment would help.

roypaulin commented on api/v1/helpers.go:
```diff
@@ -1546,6 +1546,29 @@ func (v *VerticaDB) IsSetForTLS() bool {
+// IsTLSAuthDisabledForTLSConfig returns true if tls auth is explicitly disabled for the given TLS config.
+// It does not check if tls auth is enabled/disabled for the db.
+func (v *VerticaDB) IsTLSAuthDisabledForTLSConfig(tlsConfigName string) bool {
+	disabledConfig := strings.ToLower(strings.TrimSpace(vmeta.DisableTLSAuthForConfig(v.Annotations)))
```
Isn't it better to move this to annotations.yaml?

kcmackeen replied:
I assume you mean move the strings.ToLower and strings.TrimSpace to annotations.go? That makes sense.

roypaulin reviewed (commented):
Wouldn't it be better to either have 2 annotations (enable-https-tls, enable-client-server tls) or on top of having "enable-tls-auth", add disable-https/disable-client-server?

kcmackeen commented:
> Wouldn't it be better to either have 2 annotations (enable-https-tls, enable-client-server tls) or on top of having "enable-tls-auth", add disable-https/disable-client-server?

As discussed, I changed the single string annotation to be two bool annotations (disable-https-tls-auth and disable-client-server-tls-auth). I updated the description to reflect the new annotations. And I added webhooks to prevent setting both, as well as some other invalid cases. I think it is ready for re-review.

kcmackeen commented on pkg/meta/annotations.go:
```diff
@@ -83,10 +83,11 @@ const (
+	// When that is set, these flags can be used to disable for HTTPS or Client-Server
+	// (while leaving the other enabled). They are mutually exclusive.
+	DisableTLSAuthForHTTPSAnnotation        = "vertica.com/disable-https-tls-auth"
+	DisableTLSAuthForClientServerAnnotation = "vertica.com/disable-client-server-tls-auth"
```
I have updated the code and the summary based on our discussion. Please re-review and make sure I didn't miss anything.

cchen-vertica commented on pkg/controllers/vdb/httpstlsupdate_reconciler.go:
```diff
@@ -86,8 +86,7 @@ func (h *HTTPSTLSUpdateReconciler) Reconcile(ctx context.Context, req *ctrl.Requ
 	}
 
-	if !h.Vdb.IsHTTPSConfigEnabled() ||
-		h.Vdb.IsStatusConditionTrue(vapi.HTTPSTLSConfigUpdateFinished) {
```
Why do we remove the condition check here?

kcmackeen replied:
These checks were moved into ```ShouldSkipHTTPSTLSUpdateReconcile```

cchen-vertica commented on pkg/controllers/sandbox/sandbox_controller.go:
```diff
@@ -132,7 +132,7 @@ func (r *SandboxConfigMapReconciler) Reconcile(ctx context.Context, req ctrl.Req
 		return ctrl.Result{}, err
 	}
-	prunner := cmds.MakeClusterPodRunner(log, r.Cfg, vdb.GetVerticaUser(), passwd, vmeta.UseTLSAuth(vdb.Annotations))
+	prunner := cmds.MakeClusterPodRunner(log, r.Cfg, vdb.GetVerticaUser(), passwd, vdb.IsHTTPSNMATLSAuthEnabled())
```
For pod runner, we should use IsClientServerTLSAuthEnabled()

cchen-vertica commented on pkg/controllers/vdb/verticadb_controller.go:
```diff
@@ -181,7 +181,7 @@ func (r *VerticaDBReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 		return ctrl.Result{}, err
 	}
-	prunner := cmds.MakeClusterPodRunner(log, r.Cfg, vdb.GetVerticaUser(), passwd, vmeta.UseTLSAuth(vdb.Annotations))
+	prunner := cmds.MakeClusterPodRunner(log, r.Cfg, vdb.GetVerticaUser(), passwd, vdb.IsAnyTLSAuthEnabled())
```
For pod runner, we should use IsClientServerTLSAuthEnabled()

cchen-vertica commented on pkg/controllers/vrep/replication_reconciler.go:
```diff
@@ -287,7 +286,7 @@ func (r *ReplicationReconciler) makePodFacts(ctx context.Context, vdb *vapi.Vert
 		return nil, err
 	}
-	prunner := cmds.MakeClusterPodRunner(r.Log, r.VRec.Cfg, username, password, vmeta.UseTLSAuth(vdb.Annotations))
+	prunner := cmds.MakeClusterPodRunner(r.Log, r.VRec.Cfg, username, password, vdb.IsAnyTLSAuthEnabled())
```
For pod runner, we should use IsClientServerTLSAuthEnabled()

cchen-vertica commented on pkg/controllers/vrep/replication_status_reconciler.go:
```diff
@@ -223,7 +223,7 @@ func (r *ReplicationStatusReconciler) makePodFacts(ctx context.Context, vdb *vap
 		return nil, err
 	}
-	prunner := cmds.MakeClusterPodRunner(r.Log, r.VRec.Cfg, username, password, vmeta.UseTLSAuth(vdb.Annotations))
+	prunner := cmds.MakeClusterPodRunner(r.Log, r.VRec.Cfg, username, password, vdb.IsAnyTLSAuthEnabled())
```
For pod runner, we should use IsClientServerTLSAuthEnabled()

cchen-vertica commented on pkg/controllers/vrep/replication_reconciler.go:
```diff
@@ -182,7 +181,7 @@ func (r *ReplicationReconciler) determineUsernameAndPassword(ctx context.Context
 	}
 
-	if r.TargetInfo.Vdb.IsSetForTLS() && *r.TargetInfo.Password == "" && r.Vrep.Spec.TLSConfig == "" {
+	if r.TargetInfo.Vdb.IsAnyTLSAuthEnabledWithMinVersion() && *r.TargetInfo.Password == "" && r.Vrep.Spec.TLSConfig == "" {
```
For replication, only clientserver TLS is used so we can check if clientserver TLS is enabled.

cchen-vertica commented on pkg/controllers/vdb/tlsservercertgen_reconciler.go:
```diff
@@ -85,16 +83,12 @@ func (h *TLSServerCertGenReconciler) reconcileSecrets(ctx context.Context) error
+		h.Log.Info("Reconciling TLS secret", "TLS secret", secretFieldName, "secretName", secretName)
 		// when nma secret is not empty, we can assign it to https and client TLS
-		if h.Vdb.Spec.NMATLSSecret != "" && (secretFieldName != nmaTLSSecret && secretName == "") {
+		if h.Vdb.Spec.NMATLSSecret != "" && !h.Vdb.IsAnyTLSAuthEnabled() && (secretFieldName != nmaTLSSecret && secretName == "") {
```
Why do we need `!h.Vdb.IsAnyTLSAuthEnabled()` here? When TLS is disabled, we shouldn't copy NMA secret to the other two.

kcmackeen replied:
The original purpose of this code was that, if you enable TLS and you already have a secret in nmaTLSSecret, we will copy it to httpsNMATLS.secret and clientServerTLS.secret, instead generating new ones.

The problem is the case where httpsNMATLS is disabled and clientServerTLS is enabled. Since httpsNMATLS is disabled, we will first generate an NMA secret. It will then copy the NMA secret into the clientServerTLS.secret. This seemed illogical to me; I feel like it should generate a separate secret for clientServer.

But, you are right; ```!IsAnyTLSAuthEnabled``` is wrong here. I feel like we should change the logic so that we only apply NMA secret to httpsNMATLS.secret; for clientServerTLS, I feel like we should always generate a new one. It is a change in functionality, but it makes more sense to me, given that the two TLS configs are separate now.

kcmackeen commented on pkg/controllers/vdb/clientservertlsupdate_reconciler.go:
```diff
@@ -96,17 +95,19 @@ func (h *ClientServerTLSUpdateReconciler) Reconcile(ctx context.Context, req *ct
-	res, err := h.Manager.setPollingCertMetadata(ctx)
-	if verrors.IsReconcileAborted(res, err) {
-		return res, err
+	if h.Vdb.IsHTTPSNMATLSAuthEnabled() {
```
The next line is:

`res, err1 := h.Manager.setPollingCertMetadata(ctx)`

Which sets the certs needed for polling HTTPS. If we have clientServerTLS enabled and httpsNMATLS disabled, we do not need these polling certs. This is something that was not possible before, which is why we didn't need the check.

kcmackeen commented on api/v1/verticadb_types.go:
```diff
@@ -992,6 +992,14 @@ type TLSConfigSpec struct {
+	// If 'enabled' is not specified for any  TLS configs, and the annotation is set,
+	// the operator will continue to honor the annotation (for backwards compatibility).
+	// Otherwise, 'enabled' takes precedence.
+	Enabled *bool `json:"enabled,omitempty"`
```
I see 4 cases here:
1) TLS Config is defined and "enabled" is specified; easy case -> use enabled value
2) TLS Config is defined but "enabled" is not specified; if user set a secret/mode, they obviously want TLS auth -> TRUE
3) TLS Config is not defined and annotation is false or not set; user clearly does not want TLS -> FALSE
4) TLS Config is not defined and annotation is true; see below

This is a rare but possible case for customers. Typically, if customer sets TLS annotation but no TLS config, they want operator to auto-generate their secrets for them. So this case can occur in that window between when user sets TLS and when we finish generating secret and set in spec.

So, for this case, TLS config will be enabled if annotation is true; otherwise, it is false. In the future, we will remove the annotation and this case will be false.
